### PR TITLE
mgr/cephadm: make generate_config internal, prepare before redeploy

### PR DESCRIPTION
**Pull request by adk3798**
makes generate_config into an internal function for prepare_create and calls
prepare_create whenever a service is redeployed or reconfigured

Fixes: https://tracker.ceph.com/issues/48373

Signed-off-by: Adam King <adking@redhat.com>

Some concerns here: 

The way ports are handled in the binary on redeploys here
is to  just assume the ports were checked on initial deploy and
skip the check (in code comment explains a bit more).

No prep functions are called before redploy/reconfig for osds. The
osd service has a different preparatory code path than the other 
services and I'm not sure if any of it needs to happen on a 
redeploy/reconfig. I've left them out since it seems safer and 
redeploying the osd service seemed to run successfully during
local tests with these changes.

Redeploy/reconfig won't work for custom containers, but it doesn't 
seem like it was possible before this change either.

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/pybind/mgr/cephadm/module.py [hunk 1]
```diff
@@ -1555,8 +1555,24 @@ def _daemon_action(self, daemon_type: str, daemon_id: str, host: str, action: st
                 self.mgr_service.fail_over()
                 return ''  # unreachable
             # stop, recreate the container+unit, then restart
+            try:
```
Comment (jmolmo) [inline src/pybind/mgr/cephadm/module.py:1558]: Repeating code ... what about?:

```python
....
    # Special case: redeploying the manager daemon
    if self.daemon_is_self(daemon_type, daemon_id):
        if action == 'redeploy':
            self.mgr_service.fail_over()
            return ''  # unreachable

    # Special case: redeploy/reconfig
    if action == 'redeploy' or action == 'reconfig':
        if daemon_type != 'osd':
            try:
                daemon_spec = self.cephadm_services[daemon_type].prepare_create(daemon_spec)
            except (RuntimeError, OrchestratorError) as e:
                msg = "Failed to {} daemon {} on {}: {}".format(
                    action, daemon_spec.name(), daemon_spec.host, e)
                self.log.info(msg)
                return msg
        return self._create_daemon(daemon_spec, reconfig=(action=='reconfig'))

    # Normal case.
    actions = {
        'start': ['reset-failed', 'start'],
        'stop': ['stop'],
        'restart': ['reset-failed', 'restart'],
    }

...
```

**Review by jmolmo — APPROVED** [review]
LGTM.
Feel free to adopt or not my suggestion about avoid repeating code in '_daemon_action' method

File: src/pybind/mgr/cephadm/services/container.py [hunk 2]
```diff
@@ -14,6 +16,14 @@ class CustomContainerService(CephadmService):
     def prepare_create(self, daemon_spec: CephadmDaemonSpec[CustomContainerSpec]) \
             -> CephadmDaemonSpec:
         assert self.TYPE == daemon_spec.daemon_type
+        if daemon_spec.spec is None:
+            # Exit here immediately because the required service
+            # spec to create a daemon is not provided. This is only
+            # provided when a service is applied via 'orch apply'
+            # command.
+            msg = "Required service specification not provided"
+            raise OrchestratorError(msg)
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/services/container.py:25]: I'm disagreeing with @liewegas here, but IMO we should simplify things and *only* allow daemons with specs.

File: src/pybind/mgr/cephadm/services/iscsi.py [hunk 3]
```diff
@@ -26,6 +26,13 @@ def config(self, spec: IscsiServiceSpec) -> None:
 
     def prepare_create(self, daemon_spec: CephadmDaemonSpec[IscsiServiceSpec]) -> CephadmDaemonSpec:
         assert self.TYPE == daemon_spec.daemon_type
+        # if spec is not attached to daemon_spec it is likely a redeploy or reconfig and
+        # spec should be in spec store
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/services/iscsi.py:30]: can we please fail hard, if we don't have a spec for iscsi?

File: src/pybind/mgr/cephadm/services/monitoring.py [hunk 4]
```diff
@@ -85,7 +86,15 @@ class AlertmanagerService(CephadmService):
 
     def prepare_create(self, daemon_spec: CephadmDaemonSpec[AlertManagerSpec]) -> CephadmDaemonSpec:
         assert self.TYPE == daemon_spec.daemon_type
+        # if spec is not attached to daemon_spec it is likely a redeploy or reconfig and
+        # spec should be in spec store
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/services/monitoring.py:90]: can we please fail hard, if we don't have a spec?

File: src/pybind/mgr/cephadm/services/nfs.py [hunk 5]
```diff
@@ -29,12 +29,21 @@ def config(self, spec: NFSServiceSpec) -> None:
 
     def prepare_create(self, daemon_spec: CephadmDaemonSpec[NFSServiceSpec]) -> CephadmDaemonSpec:
         assert self.TYPE == daemon_spec.daemon_type
+        # if spec is not attached to daemon_spec it is likely a redeploy or reconfig and
+        # spec should be in spec store
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/services/nfs.py:33]: can we please fail hard, if we don't have a spec here?

**Review by sebastian-philipp — CHANGES_REQUESTED** [review]

File: src/pybind/mgr/cephadm/services/nfs.py (hunk 5 above)
Comment (sebastian-philipp) [reply src/pybind/mgr/cephadm/services/nfs.py:33]: I mean, we can't properly create the spec out of thin air.

**Comment by adk3798** [comment]
This should work with _create_daemon in serve.py and the ha-rgw service present now

File: src/pybind/mgr/cephadm/services/iscsi.py (hunk 3 above)
Comment (adk3798) [reply src/pybind/mgr/cephadm/services/iscsi.py:30]: can you clarify what you mean by fail hard here (and in similar comments)? If we can't find the spec the `assert daemon_spec.spec` will fail a few lines later. Do you want to fail in a different way?

File: src/cephadm/cephadm [hunk 6]
```diff
@@ -3903,8 +3903,18 @@ def command_deploy(ctx):
...
-        daemon_ports = list(map(int, ctx.tcp_ports.split()))
+
+    # on redeploy or reconfig, ports that are to be checked
+    # could fail check that they are not "in use" since they
+    # are being used by the previously deployed service.
+    # Is there a way to check that ports received from module
+    # have changed since past deployment (which mean they would
+    # in fact need a check)? Is there a way to know if "in use"
+    # ports are being used by the service we wish to assign them
+    # to or taken by something else?
```
Comment (sebastian-philipp) [inline src/cephadm/cephadm:3914]: https://unix.stackexchange.com/a/238847/160469

But I'm :+1: with simply disabling the port-in-use check for re{deploy|config}

File: src/pybind/mgr/cephadm/module.py [hunk 7]
```diff
@@ -1575,14 +1575,21 @@ def _daemon_action(self, daemon_type: str, daemon_id: str, host: str, action: st
...
+        if action == 'redeploy' or action == 'reconfig':
+            if daemon_type != 'osd':
+                try:
+                    daemon_spec = self.cephadm_services[daemon_type_to_service(
+                        daemon_type)].prepare_create(daemon_spec)
+                except (RuntimeError, OrchestratorError) as e:
+                    msg = "Failed to {} daemon {} on {}: {}".format(
+                        action, daemon_spec.name(), daemon_spec.host, e)
+                    self.log.info(msg)
+                    return msg
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/module.py:1591]: can you move this into a different commit and explain why this is required?

**Review by sebastian-philipp — CHANGES_REQUESTED** [review]
code looks good. would be great to have two commits

File: src/cephadm/cephadm (hunk 6 above)
Comment (adk3798) [reply src/cephadm/cephadm]: Going to stick with just not rerunning this on on redeploy/reconfig for now. Maybe something to look into later.

File: src/pybind/mgr/cephadm/module.py (hunk 7 above)
Comment (adk3798) [reply src/pybind/mgr/cephadm/module.py:1614]: @sebastian-philipp  I've put this in another commit and added some explanation to that commit message. Do you want some inline code comments as well?

**Comment by sebastian-philipp** [comment]
https://pulpito.ceph.com/swagner-2021-02-02_09:06:02-rados:cephadm-wip-swagner-testing-2021-01-29-1857-distro-basic-smithi/

File: src/pybind/mgr/cephadm/module.py [hunk 8]
```diff
@@ -1598,14 +1598,21 @@ def _daemon_action(self, daemon_type: str, daemon_id: str, host: str, action: st
...
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/module.py:1614]: Unfortunately, we have the irritating situation that we have two ways to deal with errors in cephadm:

1. return errors as strings
2. return errors as `OrchestratorError` exceptions. 


After having a close look, it turned out that caching exceptions here won't work as intended:

Let's have alook at serve.py and upgrade.py:

https://github.com/ceph/ceph/blob/29f167eecf72094c33580fd403c9c18e82f3e93a/src/pybind/mgr/cephadm/serve.py#L664-L676

https://github.com/ceph/ceph/blob/29f167eecf72094c33580fd403c9c18e82f3e93a/src/pybind/mgr/cephadm/upgrade.py#L495-L512

Have a look at those except blocks: Thy both expect `daemon_action` to raise exceptions in case of errors. 

after your PR, the upgrade.py would think that the call to `_daemon_action` would have succeeded, even though `_create_daemon` was never called! What do you think of simply removing the `try-except` block?

File: src/pybind/mgr/cephadm/services/cephadmservice.py [hunk 9]
```diff
@@ -404,6 +409,8 @@ def prepare_create(self, daemon_spec: CephadmDaemonSpec) -> CephadmDaemonSpec:
         daemon_spec.ceph_conf = extra_config
         daemon_spec.keyring = keyring
 
+        daemon_spec.final_config, daemon_spec.deps = self.generate_config(daemon_spec)
+
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/services/cephadmservice.py:413]: Calling `generate_config` from `prepare_create` makes sense to me, as they're basically the same function. What I don't understand is: why are we then **still** calling `generate_config` in  `_create_daemon()`:

https://github.com/ceph/ceph/blob/b659fef6d169e96bc6d21569be82003974a2c3a9/src/pybind/mgr/cephadm/serve.py#L768-L769

I think we should only have a **single** place where we call `generate_config` per daemon deployment.

**Review by sebastian-philipp — CHANGES_REQUESTED** [review]

File: src/pybind/mgr/cephadm/services/cephadmservice.py (hunk 9 above)
Comment (adk3798) [reply src/pybind/mgr/cephadm/services/cephadmservice.py:413]: Yeah, this should not be here. I think I originally removed it then forgot to do so again when  rebasing after _create_daemon was moved to serve.py. I'll remove it again.

File: src/pybind/mgr/cephadm/module.py (hunk 8 above)
Comment (adk3798) [reply src/pybind/mgr/cephadm/module.py:1614]: Removing the try-except block works for me. I'll make that change.

File: src/pybind/mgr/cephadm/serve.py [hunk 10]
```diff
@@ -776,9 +776,6 @@ def _create_daemon(self,
                 if haspec.keepalived_container_image:
                     image = haspec.keepalived_container_image
 
-            cephadm_config, deps = self.mgr.cephadm_services[daemon_type_to_service(daemon_spec.daemon_type)].generate_config(
-                daemon_spec)
-
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/serve.py:781]: You opted for calling `generate_config` from within `prepare_create()`. This is perfectly fine!

The problem with this change is: We have code paths into `_create_daemon` that don't call `prepare_create`, like:

https://github.com/ceph/ceph/blob/4db6aebee01d1499878f2cd776aaea027105d2f9/src/pybind/mgr/cephadm/module.py#L1600-L1616

https://github.com/ceph/ceph/blob/4db6aebee01d1499878f2cd776aaea027105d2f9/src/pybind/mgr/cephadm/module.py#L1957-L1960

which unlike `_apply_service()` do not call `prepare_create()`

https://github.com/ceph/ceph/blob/4db6aebee01d1499878f2cd776aaea027105d2f9/src/pybind/mgr/cephadm/serve.py#L560-L562

And that's the problem in https://tracker.ceph.com/issues/48373 . We have to make sure that all code paths into `_create_daemon` properly call `prepare_create()`. (or simply call `prepare_create` at this point here)

**Review by sebastian-philipp — CHANGES_REQUESTED** [review]

File: src/pybind/mgr/cephadm/serve.py (hunk 10 above)
Comment (adk3798) [reply src/pybind/mgr/cephadm/serve.py]: @sebastian-philipp I can see 4 places we call `_create_daemon`. In `_apply_service` in serve.py which already calls `prepare_create`, once in osd.py where `prepare_create` isn't needed because it's making an osd,  in `_daemon_action` in cephadm/module.py which is covered in this commit as part of this PR https://github.com/ceph/ceph/pull/38549/commits/c0353e0c1d1a2580cee358774ad6e9a9f7739829 and in `_create_daemons` in cephadm/module.py which is covered by the call to `create_func` which should hold the `prepare_create` function for whatever daemon type is being created. Am I missing any?

Comment (sebastian-philipp) [reply src/pybind/mgr/cephadm/serve.py]: > Am I missing any? 

I think you're correct

**Review by sebastian-philipp — APPROVED** [review]

**Comment by sebastian-philipp** [comment]
https://pulpito.ceph.com/swagner-2021-02-11_11:00:52-rados:cephadm-wip-swagner3-testing-2021-02-10-1322-distro-basic-smithi/

* https://tracker.ceph.com/issues/49259
* https://tracker.ceph.com/issues/49287
* https://github.com/ceph/ceph/pull/38915#discussion_r575588725

**Comment by sebastian-philipp** [comment]
follow-up PR: #39458
